### PR TITLE
Rp2040 multithreaded target support

### PR DESCRIPTION
**Pull request by sberkun**
This PR adds multithreaded support from the RP2040 platform. Builds off of the work in https://github.com/lf-lang/reactor-c/pull/291.

Corresponding PR in lingua-franca: https://github.com/lf-lang/lingua-franca/pull/2178.

Some important notes on this implementation (main differences from Abhi's implementation):
 - semaphores are used as an analog for condition variables. This works because the condition variable only needs to notify at most one other thread. Will trigger a spurious wakeup later if the other thread is not waiting on the condition variable.
 - only one other thread (in addition to the main thread) is allowed. This will be a tricky constraint for the runtime to work with (in particular, scheduling enclaves and watchdog timers will need some work), but will allow for the runtime to work with no context switches.
 
 Edit: this PR is meant as part 1 of a series. This PR is meant to be minimal changes to get the threaded runtime working. The next few PRs will be:
  - modify the runtime so that the main thread also runs a worker
  - merge atomics implementation with Zephyr
  - lf_initialize_clock return value

### CODE REVIEW

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [inline core/platform/lf_rp2040_support.c:323]: The atomics implementations are pretty much copy-pasted from Zephyr support. In a future PR, I would like to merge the two somehow. Not going to do it in this PR in case I end up accidentally breaking Zephyr support.

File: core/CMakeLists.txt
Comment (sberkun) [inline core/CMakeLists.txt:96]: I tried `if (NOT (PLATFORM_ZEPHYR OR PLATFORM_RP2040))` but it didn't work. Not sure how `PLATFORM_ZEPHYR` works here.

Comment (sberkun) [reply core/CMakeLists.txt:96]: todo: fix, same as https://github.com/lf-lang/reactor-c/blob/e8a0b97917c8ef146cd074e67aa69df43f344bb6/core/platform/CMakeLists.txt#L24C1-L24C30

File: core/platform/lf_rp2040_support.c
Comment (gundralaa) [inline core/platform/lf_rp2040_support.c:270]: Iirc since the main thread in reactor c just blocks until join completes, this would block core0 from executing anything

**Review by gundralaa — COMMENTED** [review]
Pretty cool someone’s picking this up :) The only major comment I have is enabling more than 1 worker thread. With only one, it defeats the purpose of having a threaded implementation since the main reactor-c thread does not execute reactions.

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [reply core/platform/lf_rp2040_support.c:248]: Oops, some of the [multicore examples](https://github.com/raspberrypi/pico-examples/blob/master/multicore/hello_multicore/multicore.c#L24-L25) did this so I thought that maybe returning from core1 was bad. But you're right, when core1_entry returns it goes into a [low power state](https://github.com/raspberrypi/pico-bootrom/blob/ef22cd8ede5bc007f81d7f2416b48db90f313434/bootrom/bootrom_rt0.S#L371). I'll remove this, thanks!

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [reply core/platform/lf_rp2040_support.c:244]: The [semaphore implementation](https://github.com/raspberrypi/pico-sdk/blob/master/src/common/pico_sync/sem.c) seems to be based on spinlocks, so should be fine? Is there a reason why a non-multicore compatible semaphore would exist?

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [reply core/platform/lf_rp2040_support.c:270]: Yup. I should have mentioned this in the PR description, but I plan to start another PR once this one gets merged. Basically:
 - this PR: basic multicore support, only supports one worker thread. As you said, this is kind of pointless, but it's relatively straightforward.
 - next PR: modify the runtime so that the main thread also runs a worker, instead of just waiting for the worker threads. This will allow 2 workers (one main thread, one additional thread) and finally make the multicore support useful. However, this change may potentially affect all platforms, so I want it to be a seperate PR.

File: core/platform/lf_rp2040_support.c
Comment (erlingrj) [inline core/platform/lf_rp2040_support.c:71]: Should probably check the return value. Also, `_lf_initialze_clock` should return a status value. Because it could fail. But this would require updating all the platform implementations. But it would be an improvement

File: core/platform/lf_rp2040_support.c
Comment (erlingrj) [inline core/platform/lf_rp2040_support.c:216]: Check return value

File: core/platform/lf_rp2040_support.c
Comment (erlingrj) [reply core/platform/lf_rp2040_support.c:244]: Should probably check the return value here

File: core/platform/lf_rp2040_support.c
Comment (erlingrj) [inline core/platform/lf_rp2040_support.c:266]: return value

File: core/platform/lf_rp2040_support.c
Comment (erlingrj) [inline core/platform/lf_rp2040_support.c:274]: If I remember correctly we need a recursive mutex (Pico SDK has one also). But please look closer into this.

File: core/platform/lf_rp2040_support.c
Comment (erlingrj) [reply core/platform/lf_rp2040_support.c:318]: Would be good to merge them yes. But another thought is that a scheduler using them extensively will be expensive on Zephyr/Pico and cheap on x86. Maybe we need a new scheduler for 32bit platforms without atomics? Or maybe not :)

**Review by erlingrj — CHANGES_REQUESTED** [review]
This looks good. Just some nitpicks about checking the return values from the Pico SDK. Also look into whether we need a recursive mutex. When getting threaded support for FlexPRET we ran into that issue and had to create a FlexPRET lock implementation that supported recursive calls.

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [reply core/platform/lf_rp2040_support.c:318]: Definitely should test how much this affects performance once I start benchmarking on the rp2040. A new scheduler would be exciting :)

File: core/platform/lf_rp2040_support.c
Comment (gundralaa) [reply core/platform/lf_rp2040_support.c:265]: Neat! Sounds like a plan :)

File: core/platform/lf_rp2040_support.c
Comment (gundralaa) [reply core/platform/lf_rp2040_support.c:244]: Maybe just test it out? I think this one was meant for irq notification.

Comment (sberkun) [reply core/platform/lf_rp2040_support.c]: Did some ad hoc testing on the pico (raw C code, not LF) and semaphores seem to work with both cores. LF programs seem to be working as well

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [reply core/platform/lf_rp2040_support.c:73]: Right now, `stdio_init_all()` will always return true because both UART and USB are enabled, and UART initialization is infallible. We may want to support LF programs that disable them in the future (in which case this would return false), so I left a comment about that.

File: core/platform/lf_rp2040_support.c
Comment (sberkun) [reply core/platform/lf_rp2040_support.c]: looked at mutex usages and am not sure where recursive mutex is used. Changed to recursive mutex just to be safe

**Review by gundralaa — APPROVED** [review]
looks good 👍

**Comment by sberkun** [comment]
@erlingrj just rebased onto the most recent changes in main, would be great if we could get this merged in

**Comment by lhstrh** [comment]
Whoops, @sberkun, it looks like we forgot to merge this, and it has fallen a bit behind. Sorry! Could you once again fix the merge conflicts?

**Comment by lhstrh** [comment]
We should get this merged, but there are some conflicts. Maybe @petervdonovan or @erlingrj could help?

**Comment by sberkun** [comment]
@lhstrh @erlingrj Sorry for the delay! This is once again ready for review.

File: low_level_platform/impl/src/lf_rp2040_support.c
Comment (edwardalee) [inline low_level_platform/impl/src/lf_rp2040_support.c:234]: It seems like this could be easily generalized to support NUM_CORES rather than just 2.  It's not essential to do that now, so if you don't have time, just mark this resolved and open an issue.  We would need a platform to test it on anyway.

**Review by edwardalee — APPROVED** [review]
LGTM!  One comment included that we should not take as blocking the merge.

File: low_level_platform/impl/src/lf_rp2040_support.c
Comment (sberkun) [reply low_level_platform/impl/src/lf_rp2040_support.c]: Yup, that would be a good thing to do. This would be very relevant to an stm32 platform if one were added in the future. I'll open an issue for it

**Comment by sberkun** [comment]
@erlingrj looks like this is blocked because of requested changes; could you take a final look at this and unblock it?

**Review by erlingrj — APPROVED** [review]
Looks mostly good. But I have some questions about the cond bar implementation. I think the spurious wake-up problem in pthreads is not something we need to copy.

File: low_level_platform/impl/src/lf_rp2040_support.c
Comment (sberkun) [reply low_level_platform/impl/src/lf_rp2040_support.c:245]: Correct, but there's no thread handle for thread0 so nothing can join on thread0 regardless.

File: low_level_platform/impl/src/lf_rp2040_support.c
Comment (sberkun) [reply low_level_platform/impl/src/lf_rp2040_support.c:294]: Yes, spurious wakeups would be good to avoid, but it would complicate the implementation significantly to try to avoid them (i'm imagining an array to keep track of which threads are waiting on the condition variable). I took a look at https://github.com/raspberrypi/pico-sdk/pull/1101 which seems like an ideal implementation, but it's very complex.

Typical usage is like
```
do {
  lf_cond_wait(cond);
} while (condition);
```
So spurious wakeups only cause an extra iteration of this loop, which doesn't have much affect performance-wise.

File: low_level_platform/impl/src/lf_rp2040_support.c
Comment (sberkun) [reply low_level_platform/impl/src/lf_rp2040_support.c:288]: Yup, see other comment about spurious wakeups.

File: low_level_platform/impl/src/lf_rp2040_support.c
Comment (erlingrj) [reply low_level_platform/impl/src/lf_rp2040_support.c:294]: The PR you linked to looked complicated yes, we solved it a lot easier in FlexPRET here https://github.com/pretis/flexpret/blob/master/sdk/lib/src/cond.c

**Pull request merged**
